ieee754r/fmaxmagd32.c

```diff
@@ -0,0 +1,82 @@
+/* The fmaxmag functions determine the value of their argument 
+   whose magnitude is the maximum of  the  magnitudes  of  the  arguments:
+   the value of x if |x| > |y|, y if |x| < |y|, and fmax(x,y) otherwise.
+
+   Copyright (C) 2007-2019 Free Software Foundation, Inc.
+
+   This file is part of the Decimal Floating Point C Library.
+
+   Author(s): Pedro Caldeira <caldeira@linux.ibm.com>
+
+   The Decimal Floating Point C Library is free software; you can
+   redistribute it and/or modify it under the terms of the GNU Lesser
+   General Public License version 2.1.
+
+   The Decimal Floating Point C Library is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+   warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+   the GNU Lesser General Public License version 2.1 for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License version 2.1 along with the Decimal Floating Point C Library;
+   if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+   Suite 330, Boston, MA 02111-1307 USA.
+
+   Please see libdfp/COPYING.txt for more information.  */
+
+#ifndef _DECIMAL_SIZE
+#  define _DECIMAL_SIZE 32
+#  include <decimal32.h>
+#endif
+
+#include <math.h>
+#include <ieee754r_private.h>
+
+#define FUNCTION_NAME fmaxmag
+
+#include <dfpmacro.h>
+#include "convert_helpers.h"
+#include <stdio.h>
+#include <limits.h>
+
+DEC_TYPE
+INTERNAL_FUNCTION_NAME (DEC_TYPE x, DEC_TYPE y)
+{
+
+  if( FUNC_D(__isnan) (x) != 0 && FUNC_D(__isnan) (y)) return DEC_NAN; //returning NaN
+  //If just one argument is NaN return the other argument
+  if( FUNC_D(__isnan) (x) != 0) return y; 
+  if( FUNC_D(__isnan) (y) != 0) return x;
+
+  if( FUNC_D(__isinf) (x) != 0 && FUNC_D(__isinf) (y)) return FUNC_D(__fmax) (x,y); 
+  if( FUNC_D(__isinf) (x) != 0) return x; 
+  if( FUNC_D(__isinf) (y) != 0) return y; 
+
+  long expX, expY;
+
+#ifdef INT128
+  INT128 mantissaX = FUNC_D(getmant) (x, &expX) & 0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+  INT128 mantissaY = FUNC_D(getmant) (y, &expY) & 0x7FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF;
+#else
+  long long int mantissaX = FUNC_D(getmant) (x, &expX) & (ULLONG_MAX >> 1);
+  long long int mantissaY = FUNC_D(getmant) (y, &expY) & (ULLONG_MAX >> 1);
+#endif
+  long diff = expX - expY;
+/*
+  printf("\n%ld\n", expX);
+  printf("\n%ld\n", expY);
+  printf("\n%lld\n", mantissaX);
+  printf("\n%lld\n", mantissaY);
+*/
+  if (diff > 0){
+    mantissaX *= pow(10,diff);
+  }
+  else {
+    diff *= -1;
+    mantissaY *= pow(10,diff);
+  }
+  if(mantissaX > mantissaY) return x;
```

Comment:
why not something like `fabsdN(x) < fabsdN(y) ? y : x`?

This gives the compiler the freedom to choose the appropriate underlying implementation (e.g hard or soft). I am not sure if fabsdN is inlined today, but it is quick operation.

Reply:
But I think the `fabsdN(x)` uses libdecnumber. (Or I've just sabotaged myself for this whole week)
https://github.com/libdfp/libdfp/blob/af9e5e587b36872279ab7010e6f4ea583b590061/ieee754r/fabsd32.c#L40-L55

Reply:
My suggestion is wrong when x == y. if the signs differ, I think the positive value should be returned.

Reply:
That is the fallback version. However, both BID and DPD can trivially implement this by clearing the most significant bit. I believe PPC overrides these functions in sysdeps/powerpc/dfpu.

Reply:
Nevermind.  I've found out that we have versions for s390 and powerpc. :man_facepalming: